### ISSUE TITLE
Model validation

### ISSUE THREAD

**Issue by mfproto** [gomezzz/MSMatch#45]
Thank you for publishing the code.
I'm not sure if I understood correctly by looking through the code, did you generate separate validation dataset for selection of the best model to save, and how was the accuracy/loss on the validation dataset computed (since I suppose there would be no labels provided for validation, only pseudo-labels)?

**Comment by gomezzz** [comment]
> I'm not sure if I understood correctly by looking through the code, did you generate separate validation dataset for selection of the best model to save

The model with the best performance on the validation dataset is saved in the lines follow here:
https://github.com/gomezzz/MSMatch/blob/11f6e6def794b0998373157b8027439a44b558a9/models/fixmatch/fixmatch.py#L238

> how was the accuracy/loss on the validation dataset computed (since I suppose there would be no labels provided for validation, only pseudo-labels)?

The datasets we used (EuroSAT and UCM) had labels for all images. The code automatically ignores some of the labels to have some unlabeled data. In https://github.com/gomezzz/MSMatch/blob/11f6e6def794b0998373157b8027439a44b558a9/datasets/ssl_dataset.py there are two basic datasets generated ( here https://github.com/gomezzz/MSMatch/blob/11f6e6def794b0998373157b8027439a44b558a9/datasets/ssl_dataset.py#L206 ). In the rest of the code only the labels for supervised part are accessed.

In the lines here https://github.com/gomezzz/MSMatch/blob/11f6e6def794b0998373157b8027439a44b558a9/models/fixmatch/fixmatch.py#L161 the loss is computed based on that (ignoring the labels of the unlabeled part)

**Comment by mfproto** [comment]
I was thinking on validation data in `FixMatch.evaluate`. If I didn't miss something, validation / evaluation data was generated in
https://github.com/gomezzz/MSMatch/blob/11f6e6def794b0998373157b8027439a44b558a9/train.py#L127

**Comment by mfproto** [comment]
No, I meant how was the validation dataset chosen, was it the same as the test set used for evaluation of the final accuracy, was it chosen from the labeled or unlabeled part of the training dataset...

**Comment by gomezzz** [comment]
> No, I meant how was the validation dataset chosen, was it the same as the test set used for evaluation of the final accuracy 
Yes, the test and validation dataset are the same.

> was it chosen from the labeled or unlabeled part of the training dataset...
There is no unlabeled part of the dataset per se. EuroSAT and UCM only contain labeled data. The data are input the same way (simply by evaluating the model with the image) so there is no distinction in the way the model handles labeled or unlabeled data either. 

(During training, to simulate having unlabeled data, we discard are large portion of the training labels. The thus generated unlabeled data are then used for the unsupervised training as described in the paper.)

**Comment by mfproto** [comment]
> Yes, the test and validation dataset are the same.

Yes, it seemed so but I wanted to check. I'm not sure that is the best, i.e. most realistic way to perform model validation, because the trained model is indirectly tuned to perform well on the test dataset. Shouldn't the test dataset be treated as completely unseen during training? 

If the test set is 'seen' during training, we don't get truly realistic measure of model generalization. 
Please correct me if I am wrong here (but I don't think I am :) ).

In my understanding, in semi-supervised learning additional labels shouldn't be used for model validation because in realistic scenario (dataset with small number of labeled samples) we simply don't have access to other labels. So, to get realistic estimation of model performance, test set should be completely left out during training, and validation set also shouldn't contain labels or shouldn't be used at all.

In fixmatch, again if I understood correctly, they used fixed number of iterations and EMA, and used the model obtained in this way as final model. In this way they get realistic estimation of model performance.

> There is no unlabeled part of the dataset per se. EuroSAT and UCM only contain labeled data. The data are input the same way (simply by evaluating the model with the image) so there is no distinction in the way the model handles labeled or unlabeled data either.
> 
> (During training, to simulate having unlabeled data, we discard are large portion of the training labels. The thus generated unlabeled data are then used for the unsupervised training as described in the paper.)

I understand that, no questions here.

Thanks for your answers.

**Comment by gomezzz** [comment]
> because the trained model is indirectly tuned to perform well on the test dataset. Shouldn't the test dataset be treated as completely unseen during training?
If the test set is 'seen' during training, we don't get truly realistic measure of model generalization.
Please correct me if I am wrong here (but I don't think I am :) ).

I am not sure I follow. We don't train on the test set, we perform no early stopping or similar mechanism ,which would lead to overfitting on the validation set and the computational cost of training a model (between 1-3 days) did not allow us to perform any type of hyperparameter tuning or similar that would lead to overfitting on the validation set. For UCM, e.g., we performed no parameter optimization whatsoever so this is the literally the identical result that would have been obtained if we had not evaluated the test set during training.

> In my understanding, in semi-supervised learning additional labels shouldn't be used for model validation because in realistic scenario (dataset with small number of labeled samples) we simply don't have access to other labels. So, to get realistic estimation of model performance, test set should be completely left out during training, and validation set also shouldn't contain labels or shouldn't be used at all.

If your dataset has no labels how would you report accuracy on it?

> In fixmatch, again if I understood correctly, they used fixed number of iterations and EMA, and used the model obtained in this way as final model. In this way they get realistic estimation of model performance.

Yes, this is identical to what we do. We also used a fixed number of iteration and EMA (see here https://github.com/gomezzz/MSMatch/blob/11f6e6def794b0998373157b8027439a44b558a9/models/fixmatch/fixmatch.py#L88 and here https://github.com/gomezzz/MSMatch/blob/11f6e6def794b0998373157b8027439a44b558a9/models/fixmatch )/fixmatch.py#L150

**Comment by mfproto** [comment]
Ok, but you chose the final model based on performance on validation set, which is identical to the test set, right?
Of course labels are needed for performance evaluation, but the test set with its labels is not (should not be) seen during training, unlike here. I didn't say you trained on the test set, but it was used indirectly by choosing the final model based on evaluation on the test set.
You said: "The model with the best performance on the validation dataset is saved..."
Excuse me if I'm being persistent here, I just wanted to clarify this question.

**Comment by mfproto** [comment]
Yes, (if I understood everything correctly) they didn't use validation set to select the final model, they used the final (averaged) model after a fixed number of iterations: https://github.com/google-research/fixmatch/issues/69.
